misc/ipc.c

```diff
@@ -0,0 +1,238 @@
+ */
+#define G_LOG_DOMAIN "lib  misc"
+
+// default preallocation length for ipc_contexts
+#define IPC_CONTEXTS_CAP_STEP 10
+
+int
+ipc_send (struct ipc_context *context, enum ipc_relation to, const char *msg,
+          size_t len)
+{
+  (void) to;
+  if (context == NULL)
+    return -3;
+  if (msg == NULL)
+    return -2;
+  switch (context->type)
+    {
+    case IPC_PIPE:
+      return ipc_pipe_send (context->context, msg, len);
+    };
+  return -2;
+}
+
+int
+ipc_destroy (struct ipc_context *context)
+{
+  int rc = 0;
+  if (context == NULL)
+    return -1;
+  switch (context->type)
+    {
+    case IPC_PIPE:
+      rc = ipc_pipe_destroy (context->context);
+      break;
+    }
+  if (rc > -1)
+    free (context);
+  return rc;
+}
+
+char *
+ipc_retrieve (struct ipc_context *context, enum ipc_relation from)
+{
+  (void) from;
+  if (context == NULL)
+    return NULL;
+  switch (context->type)
+    {
+    case IPC_PIPE:
+      return ipc_pipe_retrieve (context->context);
+    };
+  return NULL;
+}
+
+int
+ipc_close (struct ipc_context *context)
+{
+  int rc = -1;
+  if (context == NULL || context->closed == 1)
+    return rc;
+  switch (context->type)
+    {
+    case IPC_PIPE:
+      context->closed = 1;
+      if ((rc = ipc_pipe_close (context->context)) > -1)
+        context->closed = 1;
+    }
+  return rc;
+}
+
+struct ipc_context *
+ipc_init (enum ipc_protocol type, enum ipc_relation relation)
+{
+  struct ipc_context *ctx = NULL;
+  void *context = NULL;
+  (void) relation;
+  if ((ctx = calloc (1, sizeof (*ctx))) == NULL)
+    goto exit;
+  ctx->type = type;
+  switch (type)
+    {
+    case IPC_PIPE:
+      context = ipc_init_pipe ();
+      break;
+    }
+  if (!context)
+    goto free_exit;
+  ctx->context = context;
+  return ctx;
+
+free_exit:
+  if (ctx != NULL)
+    free (ctx);
+exit:
+  return NULL;
+}
+
+struct ipc_context *
+ipc_exec_as_process (enum ipc_protocol type, struct ipc_exec_context *exec_ctx)
+{
+  struct ipc_context *pctx = NULL, *cctx = NULL;
+  pid_t pid;
+  if (exec_ctx == NULL || exec_ctx->func == NULL)
+    return NULL;
+  switch (type)
+    {
+    case IPC_PIPE:
+      if ((pctx = ipc_init (type, IPC_MAIN)) == NULL)
+        {
+          return NULL;
+        }
+    }
+
+  gvm_log_lock ();
+  pid = fork ();
+  gvm_log_unlock ();
+  if (pid < 0)
+    {
+      return NULL;
+    }
+  // we are the child process and execute given function
+  if (pid == 0)
+    {
+      if (pctx != NULL)
+        cctx = pctx;
+      else if ((cctx = ipc_init (type, IPC_CHILD)) == NULL)
+        {
+          exit (1);
+        }
+
+      if (exec_ctx->pre_func != NULL)
+        (*exec_ctx->pre_func) (cctx, exec_ctx->pre_arg);
+      (*exec_ctx->func) (cctx, exec_ctx->func_arg);
+      if (exec_ctx->post_func != NULL)
+        (*exec_ctx->post_func) (cctx, exec_ctx->pre_arg);
+      switch (type)
+        {
+        case IPC_PIPE:
+          ipc_destroy (pctx);
+          break;
+        }
+      free (exec_ctx);
```

Comment:
This is producing a lot of `free(): invalid size`

Reply:
oh, right because it is in the stack and not heap within proccess; going to fix it; thanks :)